[Refactor] Better align `from_single_file` logic with `from_pretrained` (#7496)
huggingface/diffusers: docs/source/en/api/loaders/single_file.md

specific language governing permissions and limitations under the License.
-->

# Loading Pipelines and Models via `from_single_file`

> The `from_single_file` method allows you to load supported Pipelines using a single checkpoint file as opposed to the folder format used by Diffusers. This is useful if you are working with many of the Stable Diffusion Web UI's that extensively rely on a single file to distribute all the components of a Diffusion Model.

Reviewer: ```suggestion
The `from_single_file` method allows you to load supported pipelines using a single checkpoint file as opposed to the folder format used by Diffusers. This is useful if you are working with many of the Stable Diffusion Web UI's (such as A1111) that extensively rely on a single file to distribute all the components of a diffusion model.
```

> The `from_single_file` method also supports loading models in their originally distributed format. This means that supported models that have been finetuned with other services can be loaded directly into supported Diffusers model objects and Pipelines.

Reviewer: ```suggestion
The `from_single_file` method also supports loading models in their originally distributed format. This means that supported models that have been finetuned with other services can be loaded directly into supported Diffusers model objects and pipelines.
```

> ## Pipelines that currently support `from_single_file` loading

Author: Can handle this in a follow up PR

```python
from diffusers import StableDiffusionXLPipeline, DDIMScheduler

ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"

scheduler = DDIMScheduler()
pipe = StableDiffusionXLPipeline.from_single_file(ckpt_path, scheduler=scheduler)

```

```python
from diffusers import StableDiffusionPipeline, ControlNetModel

ckpt_path = "https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/v1-5-pruned-emaonly.safetensors"

controlnet = ControlNetModel.from_pretrained("https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/v1-5-pruned-emaonly.safetensors")
pipe = StableDiffusionPipeline.from_single_file(ckpt_path, controlnet=controlnet)

```

## Loading a Model using `from_single_file`

```python
from diffusers import StableCascadeUNet

ckpt_path = "https://huggingface.co/stabilityai/stable-cascade/blob/main/stage_b_lite.safetensors"
model = StableCascadeUNet.from_single_file(ckpt_path)

```

## Override configuration options when using single file loading

> Override the default model or pipeline configuration options when using `from_single_file` by passing in the relevant arguments directly to the `from_single_file` method. Any argument that is supported by the Model or Pipeline class can be configured in this way

Reviewer: ```suggestion
Override the default model or pipeline configuration options when using `from_single_file` by passing in the relevant arguments directly to the `from_single_file` method. Any argument that is supported by the model or pipeline class can be configured in this way:
```

```python
from diffusers import StableDiffusionXLImg2ImgPipeline

ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-refiner-1.0/blob/main/sd_xl_refiner_1.0_0.9vae.safetensors"

> pipe = StableDiffusionXLPipeline.from_single_file(ckpt_path, requires_aesthetics_score=True)

Reviewer: Could we show more useful arguments here, rather, to drive the point home? I don't think `requires_aesthetics_score` isn't used that much.

```

```python
from diffusers import UNet2DConditionModel

ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"
model = UNet2DConditionModel.from_single_file(ckpt_path, device="cuda", upcast_attention=True)

```

## Downloading a single file checkpoint to a specific directory

```python
from diffusers import StableDiffusionXLPipeline

ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"
pipe = StableDiffusionXLPipeline.from_single_file(ckpt_path, local_dir="my_checkpoints")

```

## Using a Diffusers model repository to configure single file loading

Under the hood, `from_single_file` will try to determine a model repository to use to configure the components of the pipeline. You can also pass in a repository id to the `config` argument of the `from_single_file` method to explicitly set the repository to use.

```python
from diffusers import StableDiffusionXLPipeline

> ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"

Reviewer: Let's use a different SDXL-finetuned single-file format checkpoint here so that users understand that the configs aren't just limited to the original checkpoints.

repo_id = "stabilityai/stable-diffusion-xl-base-1.0"

pipe = StableDiffusionXLPipeline.from_single_file(ckpt_path, config=repo_id)

```

This can be useful in cases where model components might have been changed from what was originally distributed or in cases where a checkpoint file might not have the necessary metadata to correctly determine the configuration to use for the pipeline.

<Tip>

To learn more about how to load single file weights, see the [Load different Stable Diffusion formats](../../using-diffusers/other-formats) loading guide.

</Tip>

## Working with local files

> As of `diffusers>=0.28.0` the `from_single_file` method will attempt to configure a pipeline or model by first inferring the model type from the checkpoint file and then using the model type to determine the appropriate model repo configuration to use from the Hugging Face Hub. e.g. Any single file checkpoint based on the `StableDiffusionXL` base model will use the `stabilityai/stable-diffusion-xl-base-1.0` model repo to configure the pipeline.

Reviewer: ```suggestion
As of `diffusers>=0.28.0` the `from_single_file` method will attempt to configure a pipeline or model by first inferring the model type from the checkpoint file and then using the model type to determine the appropriate model repo configuration to use from the Hugging Face Hub. For example, any single file checkpoint based on the Stable Diffusion XL base model will use the [`stabilityai/stable-diffusion-xl-base-1.0`](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0) model repo to configure the pipeline.
```

```

<Tip>
Disabling symlinking means that the `huggingface_hub` caching mechanism has no way to determine whether a file has already been downloaded to the local directory. This means that the `hf_hub_download` and `snapshot_download` functions will download the file to the local directory every time they are called. If you are disabling symlinking, it recommended that you separate the download and loading steps to avoid downloading the same file multiple times.

</Tip>

## Using the original configuration file of a model

If you would like to use the original configuration file of a model when loading a model from a single file, you can do so with the `original_config` argument.

```python
from diffusers import StableDiffusionXLPipeline

ckpt_path = "https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/sd_xl_base_1.0_0.9vae.safetensors"
repo_id = "stabilityai/stable-diffusion-xl-base-1.0"
original_config = "https://raw.githubusercontent.com/Stability-AI/generative-models/main/configs/inference/sd_xl_base.yaml"

pipe = StableDiffusionXLPipeline.from_single_file(ckpt_path, original_config=original_config)
```

<Tip>

> When using `original_config` with local_files_only=True`, `diffusers` will attempt to infer the components based on the type signatures of pipeline class. This is not as reliable as providing a config path and might lead to errors when configuring the pipeline. Additionally, the pipeline scheduler will default to the `DDIMScheduler` if one isn't provided.

Reviewer: ```suggestion
When using `original_config` with 1local_files_only=True`, Diffusers will attempt to infer the components based on the type signatures of the underlying pipeline class. This is not as reliable as providing a config path and might lead to errors when configuring the pipeline. Additionally, the pipeline scheduler will default to the `DDIMScheduler` if one isn't provided.
```

> When using `original_config` with local_files_only=True`, `diffusers` will attempt to infer the components based on the type signatures of pipeline class. This is not as reliable as providing a config path and might lead to errors when configuring the pipeline. Additionally, the pipeline scheduler will default to the `DDIMScheduler` if one isn't provided.

Author: If `original_config` is provided we will configure the models based on it, regardless of what `local_files_only` is set to.

The interaction between `local_files_only` and `original_config` helps us determine whether or not we can fetch the pipeline `model_index.json` files and individual model `config.json` files from the hub. 

Consider this existing functionality of `from_single_file`

```python
pipe = StableDiffusionPipeline.from_single_file(
    "../checkpoints/sd21/v2-1_768-ema-pruned.safetensors",
    original_config="../checkpoints/sd21/config.yaml",
    local_files_only=True
)
```

In this case, a local original config has been provided, along with a local model and `local_files_only=True`.  Assuming the CLIP model configs have been placed correctly in the cache folder, this snippet will load even if there is no internet connection. This is because we have currently hardcoded what components we load in `from_single_file` , which is limited to the components of the Stable Diffusion pipelines, so we know what we need to load.

We can consider this "legacy" local behaviour. In this situation that we assume we shouldn't attempt to fetch anything from the internet, or it might break existing code if there is no internet connection. Admittedly, this is probably an edge case, but it would be nicer to not try to break anything. This is why in the updated code in this PR, we make use of the pipeline signature keys. 

We could avoid this by always trying to load a model_index.json file regardless of what `local_files_only` is set to, but I think the expectation from a user is that we won't be connecting to the internet.
